docs/dev/query/msq/tree-lifecycle.md

```diff
@@ -0,0 +1,99 @@
+For example, the referenced tables, attributes or functions may not exist or some expression may not be well typed.
+
+## Semantic validation
+
+The `SqlNodeAndOptions` is passed to `QueryEnvironment.validate`, which uses Calcite rules and the Pinot schema and
+function registries to validate the query.
+At this point is where we check that the tables and columns exist, that the functions are well typed, etc.
+
+## Relational algebra
+
+Once the query is validated, the `SqlNode` is converted to a `RelNode` using `QueryEnvironment.toRelation` method.
+This is done by the `SqlToRelConverter` class, which is a Calcite class that converts `SqlNode`s to `RelNode`s.
+
+Contrary to the `SqlNode`, the `RelNode` is not bound to the SQL language.
+The `RelNode`s represent the relational algebra and it is not bound to the SQL language or its syntax.
+For example, a query in MongoDB, in Casandra or in Prometheus can be represented as a `RelNode`. 
+
+In Calcite these `RelNode`s can be very abstract, representing a query plan that is not tied to any specific execution 
+algorithm or even engine.
+This is the case of the Logical rel nodes, which are the output of the `SqlToRelConverter`.
+They say what needs to be done but not how to do it.
+Should we use a hash join or a merge join? Should we use a bitmap index or a B-tree index?
+This is not specified in the logical level.
+
+Even more, at this point the query plan wasn't actually optimized.
+Calcite applies some small optimizations here like reducing some literal expressions (for example `1 + 1` to `2`) but
+the query plan is not optimized yet..
+
+## Optimization
+
+If Apache Pinot were following the Calcite architecture, this phase would optimize the logical `RelNode`s into
+physical `RelNode`s.
+But Pinot just partially follows this model.
+During optimization apply the rules defined in `PinotQueryRuleSets`, which transform the `RelNode`s in several ways:
+1. Apply some common relational algebra rules. For example, push down filters, prune project columns that are not used,
+etc.
+2. Apply some Pinot specific rules. For example, tries to optimize semi-joins into pipeline breakers, add exchanges
+to separate the query into multiple stages, etc.
+
+But these nodes are still not executable.
+
+## PlanNodes and DispatchableSubPlan
+
+In this phase, still in `QueryExecutor`, the `RelNode`s are converted to `PlanNode`s using `PinotLogicalQueryPlanner`.
+
+These `PlanNode`s were originally created to be able to serialize the query plan and send it to the servers, but they
+ended up being used to apply some final optimizations and are actually used to calculate how to distribute the
+stages between servers.
+Specifically in this phase we decide how many workers will execute each stage.
+These workers are distributed between the servers, so a server may execute the same stage more than once.
+
+## Serialization and deserialization
```

Comment:
Would add some text to make it clear what parts of the code are automatically generated on build time based on the .proto definitions.